keps/0022-diagnostics-bundle.md

```diff
@@ -0,0 +1,388 @@
+key in `operator.yaml`:
+
+```yaml
+diagnostics:
+  bundle:
+    resources:
+      - description: Zookeeper Configuration File
+        name: "zookeeper-configuration"
+        kind: Copy
+        spec:
+          path: 
+            - /opt/zookeeper/server.properties
+            - /opt/zookeeper/zoo.cfg
+          selector:
+            matchLabels:
+              app: zookeeper
+              heritage: kudo
+          objectRef:
+            kind: StatefulSet # Runs on ALL pods in the statefulset
+      - description: DNS information for running pod
+        name: "dns-information"
+        kind: Command
+        spec:
+          command: # Can be string or array
+            - nslookup
+            - google.com
+          selector:
+            matchLabels:
+              app: zookeeper
+              heritage: kudo
+          objectRef:
+            kind: Pod
+    filters:
+      - name: Authentication information
+        spec:
+          regex: "^host: %w+$"
+```
+
+This key is **OPTIONAL**. Default diagnostics collection will happen regardless
+of the `diagnostics.bundle` key's presence. Note, moving to a graph-based engine
+for KUDO will make selecting of resources much easier, rather than having to
+use magical strings with templates. Future iterations of this will reduce the
+complexity of selecting resources to run commands and files on.
+
+Filtering is an important part of diagnostics collection. It enables diagnostics
+to be portably sent to third parties that should not have sensitive information
+that logs and files can contain.
+
+By default, KUDO filters all resources (and custom resources) of values
+contained within the KUDO Instance's secrets. This is configurable with the
+`diagnostics.filterSecrets` key.
+
+There may be other fields that need to be filtered. To solve for this, KUDO
+introduces the `diagnostics.bundle.filters` key in `operator.yaml`, which
+contains a list of filters that files pass through before writing to disk.
+Custom filters use either a regular expression or an object reference and
+JSONPath to derive values to filter.
+
+All filtered values appear as `**FILTERED**` in relevant logs and files.
+
+### More Notes
+
+- Do we need to introduce a notion of the collector or controller manager
+  signing and/or encrypting bundles? TBD.
+
+## Resources
+
+### bundle.resources
+
+An individual bundle resource is represented as an element in the list inside of the
+`diagnostics.bundle.resources` key. Resources ALWAYS have the following keys:
+
+- **description**: The human-readable name of the file.
+- **name**: The machine-readable name of the file. This is used for both
+  references (if needed in the future) and filenames. Extension is OPTIONAL,
+  but may be useful for inferring mime types.
+- **kind**: The kind of bundle item.
+- **spec**: The attributes of a particular kind. This is different for every
+  kind.
+
+Also, specs may include an `objectRef`. It ALWAYS has the following keys:
```

Comment:
This and subsequent mentionings of the `objectRef` are outdated since we're going with the `selector`?

Reply:
removed